File: backend/app/models/query.py
"""Models for query and search functionality"""

from pydantic import BaseModel, ConfigDict, Field
from typing import List, Optional, Dict, Any
from datetime import datetime
from uuid import UUID


class Source(BaseModel):
    """Source citation for query response"""
    type: str  # "code" or "documentation"
    file_path: Optional[str] = None
    line_start: Optional[int] = None
    line_end: Optional[int] = None
    content: str
    relevance_score: float
    documentation_id: Optional[UUID] = None
    documentation_title: Optional[str] = None


class Query(BaseModel):
    """Query request model"""
    question: str
    repository_id: Optional[UUID] = None
    session_id: Optional[str] = None
    max_results: int = 10
    include_code: bool = True
    include_docs: bool = True


class QueryResponse(BaseModel):
    """Query response model"""
    model_config = ConfigDict(json_encoders={datetime: lambda v: v.isoformat()})

    answer: str
    sources: List[Source]
    confidence_score: float
    response_time_ms: int
    session_id: str
    related_questions: List[str] = Field(default_factory=list)


class QueryHistory(BaseModel):
    """Query history record"""
    model_config = ConfigDict(
        from_attributes=True,
        json_encoders={datetime: lambda v: v.isoformat()},
    )

    id: UUID
    repository_id: Optional[UUID] = None
    session_id: str
    question: str
    answer: str
    sources: List[Dict[str, Any]]
    feedback: Optional[int] = None  # -1, 0, 1
    response_time_ms: int
    created_at: datetime


class QueryFeedback(BaseModel):
    """Feedback on query response"""
    query_id: UUID
    feedback: int = Field(..., ge=-1, le=1)  # -1 (bad), 0 (neutral), 1 (good)
    comment: Optional[str] = None


class SearchQuery(BaseModel):
    """Search query model"""
    query: str
    repository_id: Optional[UUID] = None
    search_type: str = "hybrid"  # "semantic", "keyword", "hybrid"
    limit: int = 10
    filters: Optional[Dict[str, Any]] = None


class SearchResult(BaseModel):
    """Search result model"""
    id: str
    type: str  # "code" or "documentation"
    title: str
    content: str
    score: float
    metadata: Dict[str, Any]
    highlights: List[str] = Field(default_factory=list)

# Made with Bob

File: frontend/src/types/query.ts
/**
 * TypeScript types for query functionality
 */

export interface Source {
  type: 'code' | 'documentation';
  file_path?: string;
  line_start?: number;
  line_end?: number;
  content: string;
  relevance_score: number;
  documentation_id?: string;
  documentation_title?: string;
}

export interface Query {
  question: string;
  repository_id?: string;
  session_id?: string;
  max_results?: number;
  include_code?: boolean;
  include_docs?: boolean;
}

export interface QueryResponse {
  answer: string;
  sources: Source[];
  confidence_score: number;
  response_time_ms: number;
  session_id: string;
  related_questions: string[];
}

export interface QueryFeedback {
  query_id: string;
  feedback: -1 | 0 | 1;
  comment?: string;
}

export interface SearchQuery {
  query: string;
  repository_id?: string;
  search_type?: 'semantic' | 'keyword' | 'hybrid';
  limit?: number;
  filters?: Record<string, any>;
}

export interface SearchResult {
  id: string;
  type: 'code' | 'documentation';
  title: string;
  content: string;
  score: number;
  metadata: Record<string, any>;
  highlights: string[];
}

export interface Message {
  id: string;
  role: 'user' | 'assistant';
  content: string;
  sources?: Source[];
  confidence_score?: number;
  related_questions?: string[];
  timestamp: Date;
}

export interface StreamEvent {
  type: 'start' | 'status' | 'sources' | 'chunk' | 'answer' | 'done' | 'error';
  message?: string;
  content?: string;
  data?: any;
}

export interface QueryStats {
  total_queries: number;
  avg_response_time_ms: number;
  positive_feedback: number;
  negative_feedback: number;
  neutral_feedback: number;
}

// Made with Bob